include/nbl/builtin/glsl/ext/RadixSort/radix_sort.glsl

```diff
@@ -0,0 +1,69 @@
+#ifndef _NBL_GLSL_EXT_RADIXSORT_INCLUDED_
+#define _NBL_GLSL_EXT_RADIXSORT_INCLUDED_
+
+#define _NBL_GLSL_SCRATCH_SHARED_SIZE_DEFINED_ (_NBL_GLSL_WORKGROUP_SIZE_ * 2 * 4)
+shared uint scratch_shared[_NBL_GLSL_SCRATCH_SHARED_SIZE_DEFINED_];
+#define _NBL_GLSL_SCRATCH_SHARED_DEFINED_ scratch_shared
+
+#include <nbl/builtin/glsl/workgroup/ballot.glsl>
+
+#include "nbl/builtin/glsl/ext/RadixSort/parameters.glsl"
+
+#ifndef _NBL_GLSL_EXT_RADIXSORT_GET_PARAMETERS_DEFINED_
+#error "You need to define `nbl_glsl_ext_RadixSort_getParameters` and mark `_NBL_GLSL_EXT_RADIXSORT_GET_PARAMETERS_DEFINED_`!"
+#endif
+
+uint nbl_glsl_ext_RadixSort_extractDigit(in uint key)
+{
+	return (key >> nbl_glsl_ext_RadixSort_Parameters_t_getShift()) & 0xf;
+}
+
+uint nbl_glsl_ext_RadixSort_workgroupCompact(in uint digit, in bool save_local_histogram)
+{
+	uint scatter_idx = 0u;
+	uint local_histogram[NUM_BUCKETS];
+	for (int i = 0; i < NUM_BUCKETS; ++i)
+	{
+		const bool predicate = (i == digit);
+		nbl_glsl_workgroupBallot(predicate);
+		local_histogram[i] = nbl_glsl_workgroupBallotInclusiveBitCount();
+
+		if (predicate)
+			scatter_idx = local_histogram[i] - 1;
+	}
+
+	uint last_of_wg_idx = min((gl_WorkGroupID.x + 1u) * _NBL_GLSL_WORKGROUP_SIZE_ - 1u, nbl_glsl_ext_RadixSort_Parameters_t_getElementCountTotal() - 1u);
+	if (save_local_histogram && (gl_GlobalInvocationID.x == last_of_wg_idx))
+	{
+		for (int i = 0; i < NUM_BUCKETS; ++i)
+			scratch_shared[i] = local_histogram[i];
+	}
+	barrier();
+
+	return scatter_idx;
+}
+
+uvec2 nbl_glsl_ext_RadixSort_workgroupSort(in uvec2 data, in uint scatter_idx)
+{
+	if (gl_LocalInvocationIndex == 0u)
+	{
+		uint sum = 0;
+		scratch_shared[NUM_BUCKETS] = sum;
+		for (int i = 1; i < NUM_BUCKETS; ++i)
+			scratch_shared[i + NUM_BUCKETS] = (sum += scratch_shared[i - 1]);
+	}
+	barrier();
```

Comment:
relying on the stae of `scratch_shared` to be properly initialized is flaky, specially that noone can see that you reuquire "hidden inputs" by looking at your function declaration

I'd much rather you do
```glsl
uvec2 nbl_glsl_ext_RadixSort_workgroupSort(in uvec2 data, in uint scatter_idx, in uint hisotgram_element_current_thread)
{
   scratch_shared[(gl_LocalInvocationIndex+1u)&uint(_NBL_GLSL_WORKGROUP_SIZE_-1)] = hisotgram_element_current_thread;
   for (uint step=1u; step!=NUM_BUCKETS; step<<=1u) // simple stone-kogge adder
   {
      barrier();
      if (gl_LocalInvocationIndex>step && gl_LocalInvocationIndex<NUM_BUCKETS)
         scratch_shared[gl_LocalInvocationIndex] += scratch_shared[gl_LocalInvocationIndex-step];
   }
   barrier();
   scatter_idx += scratch_shared[digit];
   
   ... the rest as usual
}
```

Reply:
`uvec2 nbl_glsl_ext_RadixSort_workgroupSort(in uvec2 data, in uint scatter_idx, in uint hisotgram_element_current_thread)`

you cannot take input like that without broadcasting first (which means more `barrier`s) since the `local_histogram` is in the last thread's registers

i think i'll just remove this function entirely and inline it in `default_histogram.comp` because its only used by that file anyway

```
void main()
{
	uvec2 data = nbl_glsl_ext_RadixSort_getPaddedData(gl_GlobalInvocationID.x);
	uint digit = nbl_glsl_ext_RadixSort_extractDigit(data.x);

	uint scatter_idx = nbl_glsl_ext_RadixSort_workgroupCompact(digit);
	
       // after the call to workgroupCompact above I can assume that last thread has local histogram
       // in its registers, so I can just inline workgroupSort here
}
```

Reply:
Mergining compact with sort when doing the pass with compaction sounds ok.

Reviewing your changes, it seems that you forgot to inline the code?

Reply:
Also I'd still want you to use a cooperative stone-kogge adder instead of having 1 thread chug the prefix sum by itself